### PR TITLE
[SPARK-31186][PySpark][SQL] toPandas should not fail on duplicate column names

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

When `toPandas` API works on duplicate column names produced from operators like join, we see the error like:

```
ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().
```

This patch fixes the error in `toPandas` API.  

### Why are the changes needed?

To make `toPandas` work on dataframe with duplicate column names.

### Does this PR introduce any user-facing change?

Yes. Previously calling `toPandas` API on a dataframe with duplicate column names will fail. After this patch, it will produce correct result.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @HyukjinKwon

File: python/pyspark/sql/pandas/conversion.py [hunk 1]
```diff
@@ -132,25 +132,36 @@ def toPandas(self):
...
-            if isinstance(field.dataType, BooleanType) and pdf[field.name].isnull().any():
-                dtype[field.name] = np.object
+            if isinstance(field.dataType, IntegralType) and pandas_col.isnull().any():
+                dtype[fieldIdx] = np.float64
+            if isinstance(field.dataType, BooleanType) and pandas_col.isnull().any():
+                dtype[fieldIdx] = np.object
+
+        df = pd.DataFrame()
+        for index in range(len(dtype)):
+            t = dtype[index]
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/conversion.py]: nit: I think we can do:

```diff
-        for index in range(len(dtype)):
-            t = dtype[index]
+        for index, t in enumerate(dtype):
```

File: python/pyspark/sql/pandas/conversion.py [hunk 2]
```diff
@@ -132,25 +132,36 @@ def toPandas(self):
...
+            if t is not None:
+                series = pdf.iloc[:, index].astype(t, copy=False)
+            else:
+                series = pdf.iloc[:, index]
+            df.insert(index, self.schema[index].name, series, allow_duplicates=True)
 
-        for f, t in dtype.items():
-            pdf[f] = pdf[f].astype(t, copy=False)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/pandas/conversion.py]: @viirya, out of curiosity, doesn't it work?

```python
        for index, t in enumerate(dtype):
            if t is not None
                pdf.iloc[:, index] = pdf.iloc[:, index].astype(t, copy=False)
```

**Comment by HyukjinKwon** [comment]
Looks good. a couple of questions.

File: python/pyspark/sql/pandas/conversion.py (hunk 2 above)
Comment (viirya) [reply python/pyspark/sql/pandas/conversion.py]: Yea, not work, I'm not sure why.

File: python/pyspark/sql/pandas/conversion.py (hunk 1 above)
Comment (viirya) [reply python/pyspark/sql/pandas/conversion.py]: ok

**Review by HyukjinKwon — APPROVED** [review]
LGTM cc @BryanCutler @ueshin

File: python/pyspark/sql/pandas/conversion.py [hunk 3]
```diff
@@ -132,25 +132,35 @@ def toPandas(self):
         # Below is toPandas without Arrow optimization.
         pdf = pd.DataFrame.from_records(self.collect(), columns=self.columns)
 
-        dtype = {}
-        for field in self.schema:
+        dtype = [None] * len(self.schema)
+        for fieldIdx in range(len(self.schema)):
```
Comment (BryanCutler) [inline python/pyspark/sql/pandas/conversion.py]: better to use enumerate here?

File: python/pyspark/sql/pandas/conversion.py [hunk 4]
```diff
@@ -132,25 +132,35 @@ def toPandas(self):
...
+            if isinstance(field.dataType, BooleanType) and pandas_col.isnull().any():
+                dtype[fieldIdx] = np.object
+
+        df = pd.DataFrame()
+        for index, t in enumerate(dtype):
+            if t is not None:
+                series = pdf.iloc[:, index].astype(t, copy=False)
+            else:
+                series = pdf.iloc[:, index]
+            df.insert(index, self.schema[index].name, series, allow_duplicates=True)
```
Comment (BryanCutler) [inline python/pyspark/sql/pandas/conversion.py]: Does this make a copy of the data? Seems to go into a `make_block` method, but I can't tell for sure if that is doing an allocation

**Review by BryanCutler — COMMENTED** [review]
Looks good, just had a couple questions

File: python/pyspark/sql/pandas/conversion.py (hunk 3 above)
Comment (viirya) [reply python/pyspark/sql/pandas/conversion.py]: ok.

File: python/pyspark/sql/pandas/conversion.py (hunk 4 above)
Comment (viirya) [reply python/pyspark/sql/pandas/conversion.py]: Looks like so. `insert` calls `_sanitize_column` which makes a copy of the data. 

But `pdf.iloc[:, index] = pdf.iloc[:, index].astype(t, copy=False)` doesn't work as I replied earlier to @HyukjinKwon. Looks like whethere `iloc` returns a view or a copy, may depend on the context.

File: python/pyspark/sql/pandas/conversion.py (hunk 2 above)
Comment (viirya) [reply python/pyspark/sql/pandas/conversion.py]: As I read through related doc at Pandas, looks like whether `iloc` returns a view or a copy, depends on the context. That said, `pdf.iloc[:, index] = pdf.iloc[:, index]...` works sometimes but not always.

It explains why when I do simple test in Python shell like creating a Pandas DataFrame and reassigning like `pdf.iloc[:, index] = ...`, it works, but in added unit test it doesn't.

**Comment by HyukjinKwon** [comment]
Merged to master, and branch-3.0.

**Comment by viirya** [comment]
Thanks @HyukjinKwon @BryanCutler

**Pull request closed without merging**